File: agent-skills/skills/finance-analysis/scripts/reconciliation.py
#!/usr/bin/env python3
"""银行流水对账引擎(确定性,零网络、零付款动作)。

按"方向 + 金额到分 + 日期容差窗口"勾对:阶段一一对一精确勾对,阶段二疑似拆分/合并(1↔多
求和相等)只标注交人工、绝不自动合并。未达项按金额倒序(风险排序)。parity 见 selftest_reconciliation.py。

用法:`echo '{"bank":[...],"book":[...],"options":{"dateWindowDays":0}}' | python3 reconciliation.py`
"""
import sys
import json
import math
import re
import datetime

_DAY_MS = 86_400_000


def _round2(v):
    return math.floor(v * 100 + 0.5) / 100  # JS Math.round 半数进位


def _normalize(row, index, label):
    d = row.get("direction")
    if d not in ("in", "out"):
        raise ValueError(f"{label}第 {index + 1} 行:方向必须是 in 或 out,实际为 {d}")
    amt = row.get("amount")
    if not isinstance(amt, (int, float)) or isinstance(amt, bool) or not math.isfinite(amt) or amt <= 0:
        raise ValueError(f"{label}第 {index + 1} 行:金额必须是正数,实际为 {amt}")
    m = re.match(r"^(\d{4})-(\d{2})-(\d{2})$", str(row.get("date", "")).strip())
    if not m:
        raise ValueError(f"{label}第 {index + 1} 行:日期无法解析(应为 YYYY-MM-DD),实际为 {row.get('date')}")
    time_ms = int(datetime.datetime(int(m[1]), int(m[2]), int(m[3]), tzinfo=datetime.timezone.utc).timestamp() * 1000)
    return {**row, "index": index, "cents": math.floor(amt * 100 + 0.5), "time": time_ms, "direction": d, "amount": amt}


def reconcile(bank_input, book_input, options):
    window = options.get("dateWindowDays", 0) or 0
    if not isinstance(window, int) or isinstance(window, bool) or window < 0:
        raise ValueError(f"日期容差窗口必须是 >=0 的整数,实际为 {options.get('dateWindowDays')}")
    bank = [_normalize(r, i, "银行流水") for i, r in enumerate(bank_input)]
    book = [_normalize(r, i, "账面") for i, r in enumerate(book_input)]

    # 阶段一:一对一精确勾对(方向 + 金额到分 + 日期窗口内,取最近、同近取下标小)
    matched = []
    used_book = set()
    for b in bank:
        best, best_diff = None, float("inf")
        for k in book:
            if k["index"] in used_book or k["direction"] != b["direction"] or k["cents"] != b["cents"]:
                continue
            diff = abs(b["time"] - k["time"]) / _DAY_MS
            if diff > window:
                continue
            if diff < best_diff or (diff == best_diff and best is not None and k["index"] < best["index"]):
                best, best_diff = k, diff
        if best is not None:
            matched.append({"bank": b, "book": best, "dateDiffDays": math.floor(best_diff + 0.5)})
            used_book.add(best["index"])

    bank_unmatched = [b for b in bank if not any(m["bank"]["index"] == b["index"] for m in matched)]
    book_unmatched = [k for k in book if k["index"] not in used_book]

    # 阶段二:疑似拆分/合并(1↔多 求和相等),只标注交人工、被纳入的行移出未达清单
    needs_review = []
    consumed_bank, consumed_book = set(), set()

    def detect(singles, multi_pool, side, consumed_single, consumed_multi):
        for one in singles:
            if one["index"] in consumed_single:
                continue
            candidates = [m for m in multi_pool
                          if m["index"] not in consumed_multi and m["direction"] == one["direction"]
                          and abs(m["time"] - one["time"]) / _DAY_MS <= window]
            if len(candidates) < 2 or sum(c["cents"] for c in candidates) != one["cents"]:
                continue
            amt = f"{one['cents'] / 100:.2f}"
            note = (f"1 条银行流水(¥{amt})疑似对应 {len(candidates)} 条账面记录之和,请人工核对是否拆分入账"
                    if side == "bank" else
                    f"1 条账面记录(¥{amt})疑似对应 {len(candidates)} 条银行流水之和,请人工核对是否合并收付")
            needs_review.append({"side": side, "one": one, "many": candidates, "note": note})
            consumed_single.add(one["index"])
            for c in candidates:
                consumed_multi.add(c["index"])

    detect(bank_unmatched, book_unmatched, "bank", consumed_bank, consumed_book)
    detect(book_unmatched, bank_unmatched, "book", consumed_book, consumed_bank)
    bank_unmatched = [b for b in bank_unmatched if b["index"] not in consumed_bank]
    book_unmatched = [k for k in book_unmatched if k["index"] not in consumed_book]

    bank_only = sorted(bank_unmatched, key=lambda r: (-r["cents"], r["index"]))
    book_only = sorted(book_unmatched, key=lambda r: (-r["cents"], r["index"]))
    return {
        "matched": matched,
        "bankOnly": bank_only,
        "bookOnly": book_only,
        "needsReview": needs_review,
        "summary": {
            "bankCount": len(bank),
            "bookCount": len(book),
            "matchedCount": len(matched),
            "bankOnlyTotal": _round2(sum(r["cents"] for r in bank_only) / 100),
            "bookOnlyTotal": _round2(sum(r["cents"] for r in book_only) / 100),
            "matchedTotal": _round2(sum(m["bank"]["cents"] for m in matched) / 100),
            "balanced": not bank_only and not book_only and not needs_review,
        },
    }


def run(payload):
    try:
        return {"result": reconcile(payload.get("bank", []), payload.get("book", []), payload.get("options") or {})}
    except ValueError as e:
        return {"error": str(e)}


def main():
    try:
        payload = json.load(sys.stdin)
    except Exception as e:  # noqa: BLE001
        print(json.dumps({"error": f"输入解析失败: {e}"}, ensure_ascii=False))
        sys.exit(2)
    print(json.dumps(run(payload), ensure_ascii=False))


if __name__ == "__main__":
    main()

File: lib/domain/reconciliation.ts
// 银行流水对账:逻辑已迁到固定脚本 agent-skills/skills/finance-analysis/scripts/reconciliation.py
// (可直接调;parity 见 selftest_reconciliation.py)。本文件只留类型 + 同步 shell 包装,签名不变,工具零改动。
// 信任级别 = 复核者:只核对、只提示,绝不触发任何付款/转账。
import { execFileSync } from "node:child_process";
import path from "node:path";
import { getPythonPath, getBundledPluginRoot } from "@/lib/runtime/paths";
import { makeCalcReceipt, type CalcReceipt, type CalcStep, type CalcSource } from "./receipt";

export type ReconDirection = "in" | "out";

export type ReconInputRow = {
  date: string; // YYYY-MM-DD(或任何 Date 可解析格式)
  amount: number; // 正数金额(元),方向由 direction 决定
  direction: ReconDirection;
  description?: string | null;
  counterparty?: string | null;
};

export type ReconRow = ReconInputRow & {
  /** 在各自源数组(bank/book)中的原始下标,用于溯源 */
  index: number;
  cents: number;
  time: number; // 解析后的毫秒时间戳
};

export type ReconMatch = {
  bank: ReconRow;
  book: ReconRow;
  dateDiffDays: number;
};

export type ReconSplitMergeGroup = {
  /** "bank" = 1 条银行流水 对 多条账面;"book" = 1 条账面 对 多条银行流水 */
  side: "bank" | "book";
  one: ReconRow;
  many: ReconRow[];
  note: string;
};

export type ReconResult = {
  matched: ReconMatch[];
  /** 银行有、账上无(按金额倒序=风险排序) */
  bankOnly: ReconRow[];
  /** 账上有、银行无(按金额倒序) */
  bookOnly: ReconRow[];
  /** 疑似拆分/合并:金额能凑但笔数不一致,需人工核对(不自动匹配) */
  needsReview: ReconSplitMergeGroup[];
  summary: {
    bankCount: number;
    bookCount: number;
    matchedCount: number;
    bankOnlyTotal: number;
    bookOnlyTotal: number;
    matchedTotal: number;
    /** 三类未达项全空才算账银两平 */
    balanced: boolean;
  };
  /** 功能2: 可追溯计算回执；每笔勾对为一步，source 带行数，草稿标注需人工确认 */
  receipt: CalcReceipt;
};

export type ReconOptions = {
  /** 日期容差窗口(天),默认 0=要求同一天 */
  dateWindowDays?: number;
};

const SCRIPT = path.join(getBundledPluginRoot(), "skills", "finance-analysis", "scripts", "reconciliation.py");

/**
 * 功能2: 从对账结果构造 CalcReceipt。
 * - steps: 每笔勾对一步（银行行 ↔ 账面行 + 差额），带原始行号。
 * - source: bank/book 行数。
 * - asOf: 取所有输入中最晚日期的年月。
 * - settlementStatus 固定 "draft"（对账结果需人工确认，不是终值）。
 */
function buildReconReceipt(
  result: Omit<ReconResult, "receipt">,
  bankInput: ReconInputRow[],
  bookInput: ReconInputRow[]
): CalcReceipt {
  const steps: CalcStep[] = result.matched.map((match) => ({
    label: `银行[${match.bank.index}]↔账面[${match.book.index}]`,
    expr: `${match.bank.direction} ¥${match.bank.amount} @ ${match.bank.date} = ${match.book.direction} ¥${match.book.amount} @ ${match.book.date}`,
    inputs: {
      bankIndex: match.bank.index,
      bankDate: match.bank.date,
      bankAmount: match.bank.amount,
      bookIndex: match.book.index,
      bookDate: match.book.date,
      bookAmount: match.book.amount,
      dateDiffDays: match.dateDiffDays,
    },
    subtotal: match.bank.amount,
  }));

  const source: CalcSource[] = [
    { ref: "bank", recordCount: bankInput.length },
    { ref: "book", recordCount: bookInput.length },
  ];

  // asOf: 取所有输入日期中最晚的年月；无输入时用当月
  const allDates = [...bankInput, ...bookInput]
    .map((r) => r.date)
    .filter((d) => /^\d{4}-\d{2}-\d{2}$/.test(d))
    .sort();
  const now = new Date();
  const asOf =
    allDates.length > 0
      ? allDates[allDates.length - 1].slice(0, 7)
      : `${now.getFullYear()}-${String(now.getMonth() + 1).padStart(2, "0")}`;

  const caveats: string[] = [];
  if (!result.summary.balanced) {
    caveats.push(
      `账银未两平：银行有账无 ${result.bankOnly.length} 笔、账有银行无 ${result.bookOnly.length} 笔、疑似拆合并 ${result.needsReview.length} 组，需人工核对`
    );
  }

  return makeCalcReceipt({
    value: result.summary.matchedTotal,
    steps,
    source,
    basis: {
      caliberVersion: "reconciliation-v1",
      settlementStatus: "draft",
      asOf,
    },
    rounding: "half_up",
    ...(caveats.length > 0 ? { caveats } : {}),
  });
}

export function reconcileBankStatement(
  bankInput: ReconInputRow[],
  bookInput: ReconInputRow[],
  options: ReconOptions = {}
): ReconResult {
  let out: string;
  try {
    out = execFileSync(getPythonPath(), [SCRIPT], {
      input: JSON.stringify({ bank: bankInput, book: bookInput, options }),
      encoding: "utf-8"
    });
  } catch (e) {
    throw new Error(`对账脚本执行失败:${e instanceof Error ? e.message : String(e)}`);
  }
  const parsed = JSON.parse(out) as { result?: Omit<ReconResult, "receipt">; error?: string };
  if (parsed.error) throw new Error(parsed.error);
  if (!parsed.result) throw new Error("对账脚本无输出");
  const result = parsed.result;
  const receipt = buildReconReceipt(result, bankInput, bookInput);
  return { ...result, receipt };
}
